Closes #304 ADRS PCWG3 vignette (#306)
pharmaverse/admiralonco: vignettes/adrs_pcwg3.Rmd

---

> title: "Creating ADRS Using PCWG3 Guidelines"

Author: updated

output: 
  rmarkdown::html_vignette
vignette: >

>   %\VignetteIndexEntry{Creating ADRS with GCIG Criteria}

Author: updated

  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(admiraldev)
library(dplyr)
library(tibble)

> library(gt)

Reviewer: library(gt) still there

> Metastatic prostate cancer progression often includes measurements not fully captured by RECIST 1.1 criteria alone. Therefore, the PCWG3 guidelines extend these criteria to include composite measures combining RECIST 1.1 and PCWG3 rules.

Reviewer 2: general question do all patients should have bone lesions at baseline for prostate cancer? if not then the response will be on RECIST1.1 and PSA response? correct? I think PSA marker should indicate disease? Can we explain the RS domain little more for eg it include responses from RECIST1.1 which is for soft tissue , bone lesions responses, PSA test response where the values of soft and bone lesions are collects in TU,TR and PSA test results are in Labs.  And what are the response values for each type of responses. Overall response from these 3 types of responses are used or collected.. 2+2 rule in bone lesions etc

Author: 1.To my understanding, bone response is required to evaluate metastatic prostate cancer because 
it plays an important role in this disease, as stated in the PCWG3 guidelines. 
However, I agree that not all metastatic prostate cancer patients will have measurable bone lesions 
at the start of treatment. 
In those cases, response may be assessed using RECIST 1.1 criteria for soft tissue or PSA tumor marker levels. 
However, I’m not completely sure if these methods are enough without including bone response, 
and further review of the PCWG3 guidelines is needed to confirm this assumption.

2. Yes, we can explain RS better, I will update the vignette for responses and response values .

- ≥50% PSA decline from baseline.  

3. **Stable Disease (SD)**  
- Neither sufficient shrinkage to qualify for PR nor sufficient progression to qualify for PD. Indicates no significant change i.e      Insufficient change for CR, PR, or PD.  

4. **Progressive Disease (PD)**  
- Appearance of ≥2 new lesions confirmed on subsequent imaging scans (2+2 flare rule).  
- ≥25% PSA increase from nadir with an absolute increase of ≥2 ng/mL.  

5. **Not Evaluable (NE)**  
- Insufficient data for evaluation.  

This ordering reflects both RECIST 1.1 and PCWG3 priority interpretations for clinical analysis and 
this vignette demonstrates how to derive key analysis-ready parameters from response criteria.

# Required Packages

The examples of this vignette require the following packages.

```{r, warning=FALSE, message=FALSE}
library(admiral)
library(admiralonco)
library(admiraldev)
library(pharmaversesdtm)
library(pharmaverseadam)
library(metatools)
library(dplyr)
library(tibble)
library(gt)
```

# Programming Workflow
  
-   [Read in Data](#readdata)
-   [Pre-processing of Input Records](#input)
-   [Soft Tissue and Bone Lesion Responses](#assessments)

> -   [Bone Lesion Confirmation (`2+2 Flare Rule`)](#boneconfirmation)

Reviewer 2: are we doing any confirmation in the code for 2+2?

Author: We are not doing any confirmation in the code, with aviable data it is not possible , we can update data to do the same, I will update the text to remove confirmation and replace it with 2+2 rule guidance .

-   [Unconfirmed BOR](#ubor)
-   [Confirmed BOR](#cbor)
            
              
## Read in Data {#readdata}

To start, all data frames needed for the creation of `ADRS` should be read into the environment. This will be a company specific process. Some of the data frames needed are `ADSL` and `RS`.
            
 For example purpose, the SDTM and ADaM datasets (based on CDISC Pilot test data)---which are included in `{pharmaversesdtm}` and `{pharmaverseadam}`---are used.
              

> The dataset `rs_onco_pcwg3` is expected to already be available, containing soft tissue (`RECIST 1.1`) bone response etc. This vignette assumes the full dataset is preprocessed and ready for analysis.

Author: updated

```

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(
  rs,
   display_vars = exprs(USUBJID, RSTESTCD, RSSTRESC, VISIT,VISITNUM,RSDTC)
)
```

At this step, it may be useful to join `ADSL` to your `RS` domain. Only the `ADSL` variables used for derivations are selected 
at this step.

```{r eval=TRUE}
adsl_vars <- exprs(RANDDT, TRTSDT)
adrs <- derive_vars_merged(
  rs,
  dataset_add = adsl,
  new_vars = adsl_vars,
  by_vars = get_admiral_option("subject_keys")
)
```

## Pre-processing of Input Records {#input}

The next step is to assign parameter level values such as PARAMCD, PARAM,PARAMN, PARCAT1, etc. 
For this, a lookup can be created based on the SDTM RSCAT, RSTESTCD and RSEVAL values to join to the source data.

```{r, eval=TRUE, include=TRUE, message=FALSE}
param_lookup <- tribble(
  ~RSCAT, ~RSTESTCD, ~PARAMCD, ~PARAM, ~PARAMN, ~PARCAT1, ~PARCAT1N,
  
  # Soft Tissue Response (RECIST 1.1)
  "RECIST 1.1", "SFTSRESP", "SFTSRESP", "Soft Tissue Response by Investigator", 1, "RECIST 1.1", 1,
  
  # Bone Response (PCWG3)
  "PCWG SCHER PROSTATE CANCER 2016", "BONERESP", "BONERESP", "Bone Response by Investigator", 2, "PCWG3", 2,
  
  # Combined Overall Tumor Response

>   "PCWG3 and RECIST 1.1 Modified", "OVRLRESPC", "OVRLRESPC", "Combined Overall Tumor Response by Investigator-Derived", 3, "PCWG3 and RECIST 1.1 Modified", 3,

Author: Sure, I have updated PARAMCD, I have used modified because we have derived it basedPCWG3 and RECISIT .1.1 as per table, but to avoid confusion I will use word Derived

  
  # Tumor Marker Response (PCWG3)
  "PCWG SCHER PROSTATE CANCER 2016", "TMRESP", "TMRESP", "Tumor Marker Response by Investigator", 4, "PCWG3", 4,
  

>   # Overall Tumor Response (PCWG3)

Reviewer 2: this overall response includes all 3 correct?

Author: Yes, as collected in data, our assumption is that it is from all 3 responses.

  "PCWG SCHER PROSTATE CANCER 2016", "OVRLRESP", "OVRLRESP", "Overall Tumor Response by Investigator CRF", 5, "PCWG3 Collected", 5
)

```  

### Table : Overall Time Point Response

> 

Reviewer 3: Could we add a link to the source of this table? I think it is from https://www.lexjansen.com/pharmasug/2024/DS/PharmaSUG-2024-DS-287.pdf.

```{r, eval=TRUE, include=TRUE, message=FALSE}

overall_tpr_table <- tribble(
  ~`Soft Tissue (RECIST 1.1) TPR`, ~`Bone Lesion (PCWG3) TPR`, ~`Overall PCWG TPR`,
  "PD", "Any", "PD",
  "Any", "PD", "PD",
  "NE", "Non-PD, PDu, NED, or NE", "NE",
  "NED", "Non-PD", "Non-CR/Non-PD",
  "NED", "PDu", "PDu",
  "NED", "NED", "NE",
  "NED", "NE", "NE",
  "SD", "Non-PD, PDu, NED, or NE", "SD",
  "Non-CR/Non-PD", "Non-PD, PDu, NED, or NE", "Non-CR/Non-PD",
  "PR", "Non-PD, PDu, NED, or NE", "PR",
  "CR", "Non-PD, PDu, or NE", "PR (1)",
  "CR", "NED", "CR"
)

# Render Colorful Table with Fixed Footnotes
overall_tpr_table %>%
  gt() %>%
  tab_header(
    title = "Table 1: Overall Time Point Response",
    subtitle = "Soft Tissue (RECIST 1.1) TPR, Bone Lesion (PCWG3) TPR, and PCWG Combined TPR"
  ) %>%
  cols_label(
    `Soft Tissue (RECIST 1.1) TPR` = "Soft Tissue (RECIST 1.1)",
    `Bone Lesion (PCWG3) TPR` = "Bone Lesion (PCWG3)",
    `Overall PCWG TPR` = "Overall PCWG"
  ) %>%

>   tab_footnote(

Reviewer 2: just want to confirm how we are referencing these footnotes in the table?

    footnote = "When no target and non-target lesions are identified at baseline, and no new lesions are identified on-study, the response will be No Evidence of Disease (NED).",
    locations = cells_column_labels(columns = `Overall PCWG TPR`)  # Correct column name
  ) %>%
  tab_footnote(
    footnote = "Progressive Disease Unconfirmed (PDu): Temporary marker of possible PD where at least 2 new bone lesions are present, but an additional scan is required for confirmation.",
    locations = cells_column_labels(columns = `Bone Lesion (PCWG3) TPR`)  # Correct column name
  ) %>%
  tab_footnote(
    footnote = "(1) The overall TPR will be PR if target lesions were present at screening.",
    locations = cells_body(columns = `Overall PCWG TPR`, rows = 11)  # Correct column name
  ) %>%
  tab_footnote(
    footnote = "(2) The overall TPR will be Non-CR/Non-PD if no target lesions were present at screening.",
    locations = cells_body(columns = `Overall PCWG TPR`, rows = 12)  # Correct column name
  )
```

> ### Derive Combined Overall Response Records per PCWG3 Guidelines as per above table.

Reviewer 2: this is per timepoint correct? the overall response provided in the sdtm correct?

```{r, eval=TRUE, include=TRUE, message=FALSE}
soft_tpr <- adrs %>%
  filter(RSTESTCD == "SFTSRESP") %>%
  mutate(AVALC = RSSTRESC
  )
  
``` 

```{r, eval=TRUE, include=TRUE, message=FALSE}
bone_tpr <- adrs %>%
  filter(RSTESTCD == "BONERESP") %>%
  mutate(AVALC = RSSTRESC
  )
```  

```{r, eval=TRUE, include=TRUE, message=FALSE}

> overall_tpr <- soft_tpr %>%

Reviewer 2: if we derive this overall response , then we need to test this code on some actual data? also is this can be put as a function?

      AVALC %in% c("NE") & BONE_AVALC %in% c("NE") ~ "NE",        # Not Evaluable
      TRUE ~ NA_character_
    ),
    RSCAT="PCWG3 and RECIST 1.1 Modified",
    RSTESTCD="OVRLRESPC",
    RSTEST="Combined Overall Tumor Response by Investigator-Derived",
    RSSTRESC=AVALC

  ) %>%
  derive_vars_merged_lookup(
    dataset_add = param_lookup,
    by_vars = exprs(RSCAT, RSTESTCD)  # Match on RSCAT and RSTESTCD
  )

```

#Append  overall_tpr as rows into the primary analysis dataset adrs
```{r, eval=TRUE, include=TRUE, message=FALSE}
adrs <- bind_rows(
  adrs %>% mutate(AVALC=RSSTRESC) %>%
    derive_vars_merged_lookup(
      dataset_add = param_lookup,
      by_vars = exprs(RSCAT, RSTESTCD)  # Match on RSCAT and RSTESTCD
    ),
  overall_tpr %>% select(everything())  # Keep all variables
)
```

### Derive `AVAL`

Since the Prostate Cancer Working Group 3 (PCWG3) guidelines align closely with RECIST 1.1 categories for evaluating soft tissue, bone lesions, and overall disease progression, we can use the admiralonco::aval_resp() function to assign AVAL (numeric values ordered from best to worst response)

```{r}

> adrs <- adrs %>%

Reviewer 2: not sure as we also need to provide aval for PDu as we have used for bone lesion response. We should create AVAL for each ALALC value in RS

Author: Yes, I will remove using aval_resp and create custom function for the same

>     AVALC = case_when(

Reviewer: Hi @kiran-peddamudium @amitjaingsk @bundfussr. 
Below I paste a table where you can see which cases are not included in `SDTM.RS`, and also not in the derivation of the combined parameter.
Some cases are partially included -> the green color tells you what we have in the source data, and therefore in the derivation.
There is one case that we have in the data, but not in the derivation, that is subject 1148 with recist1.1 PR and bone response as NE. For him, the derivation assigns NA.
I think it's okay not to have all cases in the data, but the derivation should include all cases.

![image](https://github.com/user-attachments/assets/067e8d7e-53ba-467d-8118-8b23687cc891)

>     AVALC = case_when(

Reviewer 3: @starosto , the case Soft tissue: CR and Bone Lesion: Non-PD, PDu or NE is included twice with different overall response.

>     AVALC = case_when(

Reviewer: @bundfussr based on PHUSE article https://www.lexjansen.com/pharmasug/2024/DS/PharmaSUG-2024-DS-287.pdf
(1) The overall TPR will be PR if target lesions were present at screening. 
(2) The overall TPR will be Non-CR/Non-PD if no target lesions were present at screening. 

Sorry I did not mark this in the table

  display_vars = exprs(USUBJID, VISIT, RSCAT, RSTESTCD, RSEVAL, PARAMCD, PARAM, PARCAT1,AVALC,AVAL)
)
```

### Partial Date Imputation and Deriving `ADT`, `ADTF`, `AVISIT`, `AVISITN` etc.

If your data collection allows for partial dates, you could apply a company-specific imputation rule at this stage when deriving `ADT`. For this example, here we impute missing day to last possible date.

```{r}
adrs <- adrs %>%
  derive_vars_dt(
    dtc = RSDTC,
    new_vars_prefix = "A",
    highest_imputation = "D",
    date_imputation = "last"
  ) %>%
  derive_vars_dy(
    reference_date = TRTSDT,
    source_vars = exprs(ADT)
  ) %>%
  derive_vars_dtm(
    dtc = RSDTC,
    new_vars_prefix = "A",
    highest_imputation = "D",
    date_imputation = "last",
    flag_imputation = "time"
  ) %>%
  mutate(AVISIT = VISIT,
         AVISITN=VISITNUM)
```

#Derive ASEQ
```{r}

adrs <- derive_var_obs_number(
  dataset = adrs,
  by_vars = exprs(STUDYID, USUBJID),
  order = exprs(PARAMCD, AVISITN,ADT,AVALC),
  new_var = ASEQ
)

```

```{r, echo=FALSE}
dataset_vignette(
  adrs,
  display_vars = exprs(USUBJID, PARAMCD, ASEQ, AVISIT,AVISITN,ADT)
)
```

> ###Bone Lesion Confirmation (`The 2+2 flare Rule in PCWG3`)

Reviewer 2: nothing is derived here for 2+2 , so we can add the information at the top to describe the logic

The "2+2 rule" is a critical concept used to assess disease progression in metastatic prostate cancer, particularly when evaluating bone scans. It helps differentiate true progression from a temporary treatment-related flare reaction.

#Under the 2+2 rule:

If 2 or more new lesions appear on a bone scan after starting treatment, this might suggest disease progression—but these lesions could also result from a flare reaction, a temporary effect caused by the treatment.

To confirm progression, a follow-up scan must show 2 additional new lesions. Only then is progression definitively confirmed.

By requiring confirmation through a second scan, the 2+2 rule avoids misinterpretation of temporary changes and ensures doctors make accurate decisions about continuing or modifying treatment.

> ```{r, eval=TRUE, include=TRUE, message=FALSE}

Reviewer 2: @kiran-peddamudium Not sure if you missed these above comments?

prioritizing responses hierarchically (CR > PR > SD > PD > NE).In this part of the vignette, we will derive Combined Best Unconfirmed Overall Response based on combined response (PARAMCD = "OVRLRESPC") as Derived above.

```{r, eval=TRUE, include=TRUE, message=FALSE}
adrs <- adrs %>%
  derive_extreme_event(
    by_vars = exprs(USUBJID),                                # Derive UBOR per subject
    order = exprs(PARAMCD == "OVRLRESPC",ASEQ,AVISITN,ADT),  # Sort by overall response and chronological date
    mode = "first",                                          # Select first occurrence for UBOR
    source_datasets = list(
      adrs = adrs 
    ),
    events = list(
      bor_cr, bor_pr, bor_sd, bor_pd, bor_ne  # Event list defined above
    ),
    set_values_to = exprs(
      PARAMCD = "UBOR", 
      PARAM = "Unconfirmed Best Overall Response by Investigator", 
      PARAMN = 10       
    )
  )
```
```{r, echo=FALSE}
dataset_vignette(
  adrs%>%
    filter(PARAMCD== "UBOR"),
  display_vars = exprs(USUBJID, PARAMCD, VISIT, AVISIT, AVISITN, ADT, ADY,AVALC,AVAL)
)
```

#Derive Confirmed BOR (CBOR)

> For Confirmation ,both RECIST 1.1 and PCWG3 Guidelines Suggests the use of a 28-day confirmation period for responses like Complete Response (CR) or Partial Response (PR).

Reviewer 2: can you confirm you used the code from ADRS for RECIST1.1?

-   [Pre-processing of Input Records](#input)
-   [Soft Tissue and Bone Lesion Responses](#assessments)
-   [Derive Best Overall Response (BOR)](#bor)
-   [Derive Confirmed Best Overall Response (CBOR)](#cbor)
            
              
## Read in Data {#readdata}

To begin, all data frames needed for the creation of `ADRS` should be read into the environment. This will be a company specific process. Some of the data frames needed are `ADSL` and `RS`.
            
For demonstration purpose, the SDTM and ADaM datasets (based on CDISC Pilot test data) from `{pharmaversesdtm}` and `{pharmaverseadam}` are used.
              
In this vignette, the `RS` SDTM dataset is expected to contain:
-Soft tissue lesion responses evaluated using RECIST 1.1 criteria at each timepoint.
-Bone lesion responses assessed across visits using PCWG3 rules.
-Combined responses derived from both RECIST 1.1 (soft tissue) and PCWG3 (bone lesion) guidelines.

It is assumed that all datasets used in the example have been preprocessed and are "ready for analysis".
            
```{r message=FALSE}

# PCWG3 SDTM data
rs <- pharmaversesdtm::rs_onco_pcwg3
rs <- convert_blanks_to_na(rs)

#Exclude PSA records

> rs <- rs %>% filter(RSTEST!="Tumor Marker Response")

Author: As discussed over call,everyone agreed to keep this

      AVALC.SFTSRESP == "PR" & AVALC.BONERESP %in% c("Non-PD", "PDu", "NED", "NE") ~ "PR",

      # Scenario 10: Soft Tissue = CR + Bone Lesion = Non-PD, PDu, NE -> Overall response = PR (default to PR without screening info)
      AVALC.SFTSRESP == "CR" & AVALC.BONERESP %in% c("Non-PD", "PDu", "NE") ~ "PR",

      # Scenario 11: Soft Tissue = CR + Bone Lesion = NED -> Overall response = CR
      AVALC.SFTSRESP == "CR" & AVALC.BONERESP == "NED" ~ "CR",

      # Default: If conditions are not met, assign NA
      TRUE ~ NA_character_
    ),
    PARAMCD = "OVRLRESC",
    PARAM = "Overall Tumor Response by Investigator - Derived",
    PARAMN = 4,
    PARCAT1 = "PCWG3 and RECIST 1.1"
  )
)
``` 

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(
  adrs,
   display_vars = exprs(USUBJID,PARAM,PARAMCD,PARCAT1,AVALC,AVISIT,ADT)
)
```

## Derive `AVAL` (Numeric tumor response from AVALC values)

```{r}
adrs <- adrs %>%
  mutate(
    AVAL = case_when(
      AVALC == "CR" ~ 1,            # Complete Response
      AVALC == "PR" ~ 2,            # Partial Response
      AVALC == "SD" ~ 3,            # Stable Disease
      AVALC == "NED" ~ 4,           # No Evidence of Disease

>       AVALC == "Non-CR/Non-PD" ~ 5, # Neither Complete Response nor Progressive Disease

Reviewer 3: ```suggestion
      AVALC == "NON-CR/NON-PD" ~ 5, # Neither Complete Response nor Progressive Disease
```
see comment above.

      AVALC == "NE" ~ 6,            # Not Evaluable
      AVALC == "PDu" ~ 7,           # Progressive Disease Unconfirmed
      AVALC == "PD" ~ 8,            # Progressive Disease
      TRUE ~ NA_real_               # Default for unexpected/missing AVALC values
    )
  )
```

```{r, echo=FALSE}
dataset_vignette(
  adrs %>%

>     arrange(!!!get_admiral_option("subject_keys"), PARAMN,),

Author: updated

  display_vars = exprs(USUBJID,PARCAT1,PARAMCD, PARAM,PARAMN,AVISIT,AVISITN,ADT,AVALC,AVAL)
)
```

### Best Overall Response (BOR) and Confirmed Best Overall Response (CBOR)

> BOR represents the Best Overall Responses observed during the study, reflecting valid tumor responses such as Complete Response (CR), Partial Response (PR), Stable Disease (SD), and Progressive Disease (PD). *Progressive Disease Unconfirmed (PDu)* is not included in BOR unless it is confirmed as PD based on the 2+2 flare rule, as per PCWG3 and RECIST 1.1 guidelines.

Reviewer 3: I think we should change to
```suggestion
BOR represents the Best Overall Responses observed during the study, reflecting valid tumor responses such as Complete Response (CR), Partial Response (PR), Stable Disease (SD), and Progressive Disease (PD). For BOR and CBOR we consider *Progressive Disease Unconfirmed (PDu)* as PD if it is confirmed as PD based on the 2+2 flare rule, as per PCWG3 and RECIST 1.1 guidelines or if it is the last assessment.
```
according to what we've discussed in the last meeting.

> CBOR, on the other hand, refers to the Confirmed Best Overall Response, requiring sustained responses like CR and PR to meet confirmation criteria (e.g., 28-day confirmation). If *PDu* is not confirmed due to missing follow-up data, CBOR excludes PDu. 

Reviewer 3: I think PDu shouldn't be included but handled in the same way as for BOR.

This ensures CBOR includes only validated and confirmed tumor responses, maintaining compliance with clinical guidelines.

```{r, eval=TRUE, include=TRUE, message=FALSE}

# Define events forBest Overall Response (BOR)
bor_cr <- event(
  description = "Complete Response (CR) for Best Overall Response",
  dataset_name = "adrs",  
  condition = AVALC == "CR", 
  set_values_to = exprs(AVALC = "CR")
)

bor_pr <- event(
  description = "Partial Response (PR) for Best Overall Response",
  dataset_name = "adrs", 
  condition = AVALC == "PR",
  set_values_to = exprs(AVALC = "PR")
)

bor_sd <- event(
  description = "Stable Disease (SD) for Best Overall Response",
  dataset_name = "adrs",
  condition = AVALC == "SD",
  set_values_to = exprs(AVALC = "SD")
)

> 

Reviewer 3: An event for Non-CR/Non-PD should be added.

I also wonder if we should add an event for subjects without assessments in ADRS. We have this for the other response criteria.

What do you think?

bor_pd <- event(
  description = "Progressive Disease (PD) for Best Overall Response",
  dataset_name = "adrs",

>   condition = AVALC == "PD",

Reviewer 3: ```suggestion
  condition = AVALC == "PD",
  condition = AVALC == "PD" |
    (AVALC == "PDu" & lead(AVALC) == "PD") |
    (AVALC == "PDu" & is.na(lead(AVALC))),
```
See comment at the beginning of the section.

  set_values_to = exprs(AVALC = "PD")
)

bor_ne <- event(
  description = "Not Evaluable (NE) for Best Overall Response",
  dataset_name = "adrs",
  condition = AVALC == "NE",
  set_values_to = exprs(AVALC = "NE")
)

```
## Derive Best Overall Response (BOR)

Use the defined events to derive BOR based on the first occurrence in the adrs dataset, prioritizing 
responses hierarchically (CR > PR > SD > PD > NE).

> **Progressive Disease Unconfirmed (PDu)** is excluded as it represents a provisional status and 
> can only be escalated to **Progressive Disease (PD)** if confirmed by the "2+2 flare rule."
> 

Reviewer 3: ```suggestion
```
See comment at the beginning of the section "Best Overall Response (BOR) and Confirmed Best Overall Response (CBOR)".

In this part of the vignette, we will derive Best Overall Response based on combined response 

> (PARAMCD = "OVRLRESC") as derived above.

Reviewer 3: ```suggestion
(`PARAMCD = "OVRLRESC"`) as derived above.
```

```{r, eval=TRUE, include=TRUE, message=FALSE} 

# Derive BOR

> adrs <- adrs %>%

Reviewer 2: as we added the case of uPD in the sdtm RS data uPD for 01-701-1115, this subjeft doesn't show up in the BOR as you are not considering the uPD as final response. But as we discussed , as a conservative approach we should consider it PD and mention in the text about this case. Also, we should not keep any subject Blank BOR/CBOR.

Author: Updated , Please check and resolve

Reviewer 2: cbor/bor records are there for all subjects

  derive_extreme_event(
    by_vars = exprs(USUBJID), 
    events = list(bor_cr, bor_pr, bor_sd, bor_pd, bor_ne), # Hierarchical order
    source_datasets = list(adrs = adrs %>% filter(PARAMCD == "OVRLRESC")), # Use derived responses (OVRLRESC)
    order = exprs(event_nr, ADT), # Prioritize earliest valid event
    tmp_event_nr_var = event_nr,                            
    mode = "first",  # Retain the best response observed at the first occurrence
    set_values_to = exprs(
      PARAMCD = "BOR",
      PARAM = "Best Overall Response",
      PARAMN = 5
    )
  )
```

```{r, echo=FALSE}
dataset_vignette(
  adrs%>%
  filter(PARAMCD== "BOR"),
  display_vars = exprs(USUBJID,PARAM,PARAMCD,AVISIT, AVISITN,ADT,AVALC,AVAL)
)
```

## Derive Confirmed BOR (CBOR), i.e., CR and PR

As per RECIST 1.1 and PCWG3 guidelines, Complete Response (CR) and Partial Response (PR) require confirmation within a **28-day period** to ensure their validity.

```{r, eval=TRUE, include=TRUE, message=FALSE}

> # Confirmed CR Event with 28-day persistence

Reviewer 2: @kiran-peddamudium @bundfussr for CBOR , we still need all patients records to show them as SD, PD etc. currently you created records for the subjects which only has CR,PR confirmed. we have dome this in other ADRS vignettes you can refer.. Stefan can help

Author: Updated check and resolve

cbor_cr <- event_joined(
  description = "Confirmed Complete Response (CR)",
  dataset_name = "adrs",
  join_vars = exprs(AVALC, ADT),
  join_type = "after",
  first_cond_upper = AVALC.join == "CR" & ADT.join >= ADT + 28,  # Follow-up within 28-day window
  condition = AVALC == "CR" & all(AVALC.join == "CR"),           # All linked records must also be CR
  set_values_to = exprs(AVALC = "CR")                            # Set response as Confirmed CR
)

# Confirmed PR Event with 28-day persistence
cbor_pr <- event_joined(
  description = "Confirmed Partial Response (PR)",
  dataset_name = "adrs",
  join_vars = exprs(AVALC, ADT),
  join_type = "after",
  first_cond_upper = AVALC.join %in% c("CR", "PR") & ADT.join >= ADT + 28, # Include CR as confirmation
  condition = AVALC == "PR" & all(AVALC.join %in% c("CR", "PR")),          # Ensure all events align
  set_values_to = exprs(AVALC = "PR")
)

```  

```{r, eval=TRUE, include=TRUE, message=FALSE}
adrs <- adrs %>%
  derive_extreme_event(
    by_vars = exprs(USUBJID),                 

>     events = list(cbor_cr, cbor_pr),         

Reviewer 3: ```suggestion
    events = list(cbor_cr, cbor_pr, bor_sd, bor_pd, bor_ne),         
```
to add the other responses.

Please check [Responses & Bone Lesion Confirmation Guidance (The 2+2 Flare Rule)](https://4870646.fs1.hubspotusercontent-na1.net/hubfs/4870646/KEO149_Info_PCWG3_03.2.pdf) for more details.

# Programming Workflow
  
-   [Read in Data](#readdata)
-   [Pre-processing of Input Records](#input)
-   [Derive Best Overall Response (BOR)](#bor)
-   [Derive Confirmed Best Overall Response (CBOR)](#cbor)
-   [Other Endpoints](#other)
            
              
## Read in Data {#readdata}

To begin, all data frames needed for the creation of `ADRS` should be read into the environment. This will be a company specific process. Some of the data frames needed are `ADSL` and `RS`.
            
For demonstration purpose, the SDTM and ADaM datasets (based on CDISC Pilot test data) from `{pharmaversesdtm}` and `{pharmaverseadam}` are used.
              
In this vignette, the `RS` SDTM dataset is expected to contain:

- Soft tissue lesion responses evaluated using RECIST 1.1 criteria at each timepoint.
- Bone lesion responses assessed across visits using PCWG3 rules.
- Combined responses derived from both RECIST 1.1 (soft tissue) and PCWG3 (bone lesion) guidelines as collected at source.

> It is assumed that all datasets used in the example have been preprocessed and are "ready for analysis".

Author: Deleted this line as discussed

### Progressive Disease Unconfirmed (PDu)

-   At least 2 new bone lesions have appeared _within the flare window_ compared to baseline  _or_
-   At least 2 new bone lesions have appeared _outside the flare window_ compared to the first post-treatment scan.

Note: if there is no following visit (final visit), the time point remains at PDu.

### Progressive Disease (PD)

-   At least 2 new bone lesions had appeared _within the flare window_ compared to baseline (PDu), and at least 2 additional new bone lesions have been found on the next scan confirming progression (PD). The date of progression is the date of the scan showing the first 2 lesions.  _or_
-   At least 2 new bone lesions had appeared _outside the flare window_ compared to the first post-treatment scan (PDu), and the 2 lesions are persistent on the next scan confirming progression (PD). The date of progression is the data of the scan that first documents the second lesion.

See [**Bone Lesion Confirmation Guidance (The 2+2 Rule)**](#22rule) for further details.

### Not Evaluable (NE)

> When imaging is entirely missing or was not done, but bone lesions were present at baseline.

Reviewer 3: ```suggestion
When imaging is entirely missing or was not done.
```
as discussed with Amit (https://github.com/pharmaverse/admiralonco/pull/306#discussion_r2278926054)

```{r}
adrs <- adrs %>%
  derive_vars_dtm(
    dtc = RSDTC,
    new_vars_prefix = "A",
    highest_imputation = "D",
    date_imputation = "last"
  ) %>%
  derive_vars_dtm_to_dt(exprs(ADTM)) %>%
  derive_vars_dy(
    reference_date = TRTSDT,
    source_vars = exprs(ADT)
  ) %>%
  mutate(
    AVISIT = VISIT,
    AVISITN = VISITNUM
  )
```

### Derive `PARAMCD`, `PARAM`, `PARAMN`

The next step is to assign parameter level values such as `PARAMCD`, `PARAM`, `PARAMN` to values collected from source, etc. For this, a lookup can be created based on the SDTM `RSTESTCD` values.

```{r, eval=TRUE, include=TRUE, message=FALSE}
# Prepare param_lookup for SDTM RSTESTCD to add metadata
param_lookup <- tibble::tribble(
  ~RSTESTCD,  ~PARAMCD,   ~PARAM,                                   ~PARAMN,
  "SFTSRESP", "SFTSRESP", "Soft Tissue Response by Investigator",         1,
  "BONERESP", "BONERESP", "Bone Response by Investigator",                2,
  "OVRLRESP", "OVRLRESP", "Overall Tumor Response by Investigator",       3
)

adrs <- adrs %>%

>   left_join(param_lookup, by = "RSTESTCD") %>%

Reviewer 3: ```suggestion
  derive_vars_merged_lookup(
    dataset_add = param_lookup,
    by_vars = exprs(RSTESTCD)
  ) %>%
```
as discussed (https://github.com/pharmaverse/admiralonco/pull/306#discussion_r2169334058)

    )
  )
```

```{r, echo=FALSE}
dataset_vignette(
  adrs %>%
    arrange(!!!get_admiral_option("subject_keys"), AVISITN, PARAMN ),
  display_vars = exprs(USUBJID, PARCAT1, PARAMCD, PARAM, PARAMN, AVISIT, AVISITN, ADT, AVALC, AVAL)
)
```

## Best Overall Response (BOR) and Confirmed Best Overall Response (CBOR)

BOR represents the Best Overall Responses observed during the study, reflecting valid tumor responses such as Complete Response (CR), Partial Response (PR), Stable Disease (SD), and Progressive Disease (PD).

CBOR, on the other hand, refers to the Confirmed Best Overall Response, requiring sustained responses like CR and PR to meet confirmation criteria, such as persistence over a predefined confirmation period (e.g., 28 days), based on PCWG3 guidelines.

For both BOR and CBOR, if _PDu remains the last recorded assessment_ without follow-up confirmation, it is classified as _SD_ in this vignette. However, as a more conservative approach, you may choose to classify it as _PD_. Please refer to your study documentation and protocol requirements to confirm the preferred approach to handling _unresolved PDu_. 

> Additionally, for CBOR, if _CR_ or _PR_ cannot be confirmed, it is classified as _SD_ in this vignette. Please check the event **bor_sd** defined in the next section.

Reviewer 3: Thanks, looks good. Just some minor changes:
```suggestion
Additionally, for CBOR, if _CR_ or _PR_ cannot be confirmed, it is classified as _SD_ as usually done for RECIST. Please check the event `bor_sd` defined in the next section.
```

  condition = AVALC == "PD",
  set_values_to = exprs(AVALC = "PD")
)

bor_ne <- event(
  description = "Not Evaluable (NE) ",
  dataset_name = "adrs",
  condition = AVALC == "NE",
  set_values_to = exprs(AVALC = "NE")
)
bor_ned <- event(
  description = "No Evidence of Disease (NED) ",
  dataset_name = "adrs",
  condition = AVALC == "NED",
  set_values_to = exprs(AVALC = "NED")
)

no_data_missing <- event(
  description = paste(
    "Define missing response (MISSING) for all patients in adsl (should be used",
    "as last event)"
  ),
  dataset_name = "adsl",
  condition = TRUE,
  set_values_to = exprs(AVALC = "MISSING"),
  keep_source_vars = adsl_vars
)

```

## Derive Best Overall Response (BOR) {#bor}

Use the defined events to derive BOR based on the first occurrence in the `adrs` dataset, prioritizing responses hierarchically (CR \> PR \> SD \> Non-CR/NON-PD \> PD \> NE \> NED \> MISSING).

In this part of the vignette, we will derive Best Overall Response based on combined response (`PARAMCD = "OVRLRESC"`) as derived above.

```{r, eval=TRUE, include=TRUE, message=FALSE}

# Derive BOR
adrs <- adrs %>%
  derive_extreme_event(
    by_vars = get_admiral_option("subject_keys"), 

>     events = list(bor_cr, bor_pr,bor_sd,bor_pd,bor_non_crpd,bor_ne,bor_ned,no_data_missing),

Reviewer 3: ```suggestion
    events = list(bor_cr, bor_pr, bor_sd, bor_non_crpd,bor_pd, bor_ne, bor_ned, no_data_missing),
```
`bor_pd` should be after `bor_non_crpd` to follow the hierarchy above.

```{r, eval=TRUE, include=TRUE, message=FALSE}
# Confirmed CR Event with 28-day persistence
cbor_cr <- event_joined(
  description = "Confirmed Complete Response (CR)",
  dataset_name = "adrs",
  join_vars = exprs(AVALC, ADT),
  join_type = "after",
  first_cond_upper = AVALC.join == "CR" & ADT.join >= ADT + 28, # Follow-up within 28-day window
  condition = AVALC == "CR" & all(AVALC.join == "CR"), # All linked records must also be CR
  set_values_to = exprs(AVALC = "CR") # Set response as Confirmed CR
)

# Confirmed PR Event with 28-day persistence
cbor_pr <- event_joined(
  description = "Confirmed Partial Response (PR)",
  dataset_name = "adrs",
  join_vars = exprs(AVALC, ADT),
  join_type = "after",
  first_cond_upper = AVALC.join %in% c("CR", "PR") & ADT.join >= ADT + 28, # Include CR as confirmation
  condition = AVALC == "PR" & all(AVALC.join %in% c("CR", "PR")), # Ensure no events other than CR or PR in between
  set_values_to = exprs(AVALC = "PR")
)
```

```{r, eval=TRUE, include=TRUE, message=FALSE}
adrs <- adrs %>%
  derive_extreme_event(
    by_vars = get_admiral_option("subject_keys"),                

>     events = list(cbor_cr, cbor_pr,bor_sd,bor_pd,bor_non_crpd,bor_ne,bor_ned,no_data_missing),         

Reviewer 3: ```suggestion
    events = list(cbor_cr, cbor_pr, bor_sd, bor_non_crpd, bor_pd, bor_ne, bor_ned, no_data_missing),
```
as for BOR

Author: updated

### Progressive Disease (PD)

-   At least 2 new bone lesions had appeared _within the flare window_ compared to baseline (PDu), and at least 2 additional new bone lesions have been found on the next scan confirming progression (PD). The date of progression is the date of the scan showing the first 2 lesions.  _or_
-   At least 2 new bone lesions had appeared _outside the flare window_ compared to the first post-treatment scan (PDu), and the 2 lesions are persistent on the next scan confirming progression (PD). The date of progression is the data of the scan that first documents the second lesion.

See [**Bone Lesion Confirmation Guidance (The 2+2 Rule)**](#22rule) for further details.

### Not Evaluable (NE)

When imaging is entirely missing or was not done, but bone lesions were present at baseline.

### No Evidence of Disease (NED)

No bone lesions are present on the scan (whether some were present at baseline and have completely disappeared or whether there were no bone lesions from the start).

### Non-Progressive Disease (Non-PD)

Neither PD, PDu, NED or NE.

### Bone Lesion Confirmation Guidance (The 2+2 Flare Rule) {#22rule}

**Progression on a Bone Scan:**

> Please note: Progression on a bone scan is defined based on whether the assessment occurs during the flare period or afterwards.The flare period is generally considered to be the first 8--12 weeks post-baseline, primarily involving the initial post-baseline assessment. Please refer to your study documentation to confirm how the flare period is defined.

Author: Updated

**Flare Period:** Bone Scan Progression Criteria for the First Post-Baseline Scan: If two or more new bone lesions are identified at the first post-baseline assessment compared to the baseline scan, and then at the second assessment (conducted at least 6 weeks later), two or more additional new lesions are observed (resulting in a cumulative total of four or more new lesions since the baseline scan), the progression date is set as the date of the initial scan. This process is referred to as the "2+2 rule."

**After Flare Period:** If no progression is detected at the first assessment, the initial post-baseline scan(with in flare period) becomes the new reference point for subsequent scans, If two or more new lesions appear relative to this new baseline, and their persistence or increase in number is confirmed by a follow-up scan at least 6 weeks later, the progression date is recorded as the date of the first scan that noted the appearance of the two new lesions.

#### Examples of the 2+2 Rule

> **Flare Period Progression** A patient with metastatic prostate cancer has a baseline bone scan showing 10 lesions. At the first post-baseline scan (within the flare period), 2 new lesions are identified, increasing the total count to 12. This finding is labeled as Progressive Disease Unconfirmed (PDu). A follow-up scan conducted at least 6 weeks later reveals 2 additional new lesions (totaling 4 new lesions compared to baseline). Progression is confirmed based on the 2+2 rule, and the progression date is set as the date of the first post-baseline scan. At the first post-baseline scan response is re-assigned from PDu to PD.

Author: updated

> **Flare Phenomenon** In another scenario, the same patient has a baseline bone scan showing 10 lesions. The first post-baseline scan during the flare period identifies 2 new lesions, increasing the count to 12. However, a follow-up scan conducted 6 weeks later reveals no additional new lesions. According to the 2+2 rule, this is classified as a flare phenomenon, not true progression. At the first post-baseline scan response is re-assigned from PDu to Non-PD.

Author: updated

# Required Packages

The examples of this vignette require the following packages.

```{r, warning=FALSE, message=FALSE}
library(admiral)
library(admiralonco)
library(pharmaversesdtm)
library(pharmaverseadam)
library(dplyr)
library(tibble)
```

# Programming Workflow

-   [Read in Data](#readdata)
-   [Pre-processing of Input Records](#input)
-   [Derive Best Overall Response (BOR)](#bor)
-   [Derive Confirmed Best Overall Response (CBOR)](#cbor)
-   [Other Endpoints](#other)

## Read in Data {#readdata}

To begin, all data frames needed for the creation of `ADRS` should be read into the environment. This will be a company specific process. Some of the data frames needed are `ADSL` and `RS`.

For demonstration purpose, the SDTM and ADaM datasets (based on CDISC Pilot test data) from `{pharmaversesdtm}`[^1] and `{pharmaverseadam}` are used.

[^1]: Currently the `RS` domain contains some data issues, which will be resolved in the next `{pharmaversesdtm}` release. For the purposes of this vignette, these issues have been manually corrected using the following code:

> rs <- rs%>% mutate( RSSTRESC = case_when( USUBJID == "01-701-1275" & VISIT == "WEEK 8" & RSTESTCD %in% c("BONERESP", "OVRLRESP") ~ "PD", USUBJID == "01-701-1097" & VISIT == "WEEK 16" & RSTESTCD %in% c("BONERESP", "OVRLRESP") ~ "PD", TRUE ~ RSSTRESC ))

Author: Updated

```{r, eval=TRUE, include=TRUE, message=FALSE}
# Prepare param_lookup for SDTM RSTESTCD to add metadata
param_lookup <- tibble::tribble(
  ~RSTESTCD,  ~PARAMCD,   ~PARAM,                                   ~PARAMN,
  "SFTSRESP", "SFTSRESP", "Soft Tissue Response by Investigator",         1,
  "BONERESP", "BONERESP", "Bone Response by Investigator",                2,
  "OVRLRESP", "OVRLRESP", "Overall Tumor Response by Investigator",       3
)

adrs <- adrs %>%
  left_join(param_lookup, by = "RSTESTCD") %>%
  mutate(
    PARCAT1 = RSCAT,
    AVALC = RSSTRESC
  )
```

## Derive Combined Overall Time Point Response by Investigator (`OVRLRESC`) Parameter

Although `OVRLRESP`, representing the Overall Tumor Response by Investigator is available in the source data, we have re-derived the combined overall response by Investigator (`OVRLRESC`). This derivation follows the rules from the **PCWG3 and RECIST 1.1 combined response interpretation**, as described in the PharmaSUG 2024 publication on metastatic prostate cancer response criteria ([PharmaSUG 2024, DS-287](https://www.lexjansen.com/pharmasug/2024/DS/PharmaSUG-2024-DS-287.pdf)).

> 

Reviewer 3: @amitjaingsk , @kiran-peddamudium , I think we have an issue with the derivation of the combined response. Consider for example subject 01-701-1097. It has PDu at WEEK 16 and PD at WEEK 24 (for BONERESP). As SFTSRESP is PR and PR, we derive the combined response as PR and PD according to the table from the PharmaSUG paper. However, the paper assumes that the bone lesion response is updated if the next scan is available. I.e., BONERESP at WEEK 16 should be PD and then the combined response should be PD as well.

Do you agree?

Reviewer: Both options are good. Creating a new parameter would help to understand the process, while updating the `SDTM.RS` is better for the accuracy of data collection.

Author: I am considering creating a new variable, PD_CONFIRMED, as a flag (Y/N) to help transition PDu to PD/Non-PD while deriving OVRLRESC (Derived). This approach will clearly distinguish between OVRLRESP (collected) and OVRLRESC (derived) while maintaining traceability. Do you agree?"

Author: Creating new parameter is fine, but can I try creating Flag like PD_CONFIRM="Y" for records like this and then use it while deriving OVRLRESC , we can add text distinguishing it from OVRLRESP (collected) , also quote footnotes to validate transition from PDu to PD while deriving OVRLRESC?

Reviewer 3: Ah yes, the value for OVRLRESP at WEEK 16 is also wrong.

Maybe we should
- silently update the SDTM after reading in RS such that PDu for BONERESP is replaced unless it is the last assessment and
- remove the OVRLRESP records.

Then we don't need to explain the differences between OVRLRESP and OVRLRESC. (It would also be hard to explain why the OVRLRESP values are collected in a way such that you can't use them.)
In the next release we could update BONERESP and OVRLRESP in pharmaversesdtm and remove the update of BONERESP in the vignette and display the OVRLRESP records in the vignette.

@amitjaingsk , @kiran-peddamudium , @starosto , what do you think?

Reviewer 2: @kiran-peddamudium @bundfussr, its fine with me Stefan to fix the RS early on.. Kiran I think you updating the data in thelater in the program. Can you adjust RS directly after reading from RAW when you remove PSA record

Author: Updated

> Although `OVRLRESP`, representing the Overall Tumor Response by Investigator is available in the source data, we have re-derived the combined overall response by Investigator (`OVRLRESC`). This derivation follows the rules from the **PCWG3 and RECIST 1.1 combined response interpretation**, as described in the PharmaSUG 2024 publication on metastatic prostate cancer response criteria ([PharmaSUG 2024, DS-287](https://www.lexjansen.com/pharmasug/2024/DS/PharmaSUG-2024-DS-287.pdf)).

Author: updated

      # Scenario 11: Soft Tissue = CR + Bone Lesion = NON-PD, PDu, NE -> Overall response = PR
      # ((1) The overall TPR will be PR if target lesions were present at screening.)
      AVALC.SFTSRESP == "CR" & AVALC.BONERESP %in% c("NON-PD", "PDu", "NE") ~ "PR",

      # Soft Tissue = CR + Bone Lesion = NON-PD, PDu, NE -> Overall response =Non-CR/NON-PD
      # (2) The overall TPR will be Non-CR/NON-PD if no target lesions were present at screening.)
      # AVALC.SFTSRESP == "CR" & AVALC.BONERESP %in% c("NON-PD", "PDu", "NE") ~ "Non-CR/NON-PD",

      # Scenario 12: Soft Tissue = CR + Bone Lesion = NED -> Overall response = CR
      AVALC.SFTSRESP == "CR" & AVALC.BONERESP == "NED" ~ "CR",

      # Default: If conditions are not met, assign NA
      TRUE ~ NA_character_
    ),
    PARAMCD = "OVRLRESC",
    PARAM = "Overall Tumor Response by Investigator - Derived",
    PARAMN = 4,
    PARCAT1 = "PCWG3 and RECIST 1.1"
  )
)
``` 

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(
  adrs,
  display_vars = exprs(USUBJID, PARAM, PARAMCD, PARCAT1, AVALC, AVISIT, ADT)
)
```

## Derive `AVAL` (Numeric tumor response from `AVALC` values)

> 

Author: Updated

      AVALC == "NE" ~ 8, # Not Evaluable
      AVALC == "NED" ~ 9, # No Evidence of Disease
      TRUE ~ NA_real_ # Default for unexpected/missing AVALC values
    )
  )
```

```{r, echo=FALSE}
dataset_vignette(
  adrs %>%
    arrange(!!!get_admiral_option("subject_keys"), AVISITN, PARAMN, ),
  display_vars = exprs(USUBJID, PARCAT1, PARAMCD, PARAM, PARAMN, AVISIT, AVISITN, ADT, AVALC, AVAL)
)
```

### Create last assessment flag for OVRLRESC for dealing with last recorded assessment `Pdu`

```{r}
adrs <- adrs %>%
  restrict_derivation(
    derivation = derive_var_extreme_flag,
    args = params(
      by_vars = exprs(USUBJID), # Group records by Subject ID
      order = exprs(ADT), # Sort records by Assessment Date
      new_var = LAST_REC_FL, # Name of the derived flag for last record
      mode = "last" # Flag only the last record per group
    ),
    filter = PARAMCD == "OVRLRESC" # Restriction: Derive flag only for OVRLRESC records
  )
```

```{r, echo=FALSE}
dataset_vignette(
  adrs %>%
    filter(PARAMCD == "OVRLRESC" & LAST_REC_FL == "Y"),
  display_vars = exprs(USUBJID, PARAMCD, AVISIT, AVISITN, ADT, AVALC, AVAL, LAST_REC_FL)
)
```

> 

Reviewer 3: Using
```
bor_sd <- event(
  description = "Stable Disease (SD) ",
  dataset_name = "adrs",
  condition = AVALC %in% c("CR", "PR", "SD", "PDu"),
  set_values_to = exprs(AVALC = "SD")
)
```
instead of deriving `LAST_REC_FL` and defining the additional `pdu_to_sd_event` looks clearer and simpler to me.

Using two events for the same response value (SD) only works in this special case because PDu can occur only at the last assessment. In general this could result in incorrect dates.

Both approaches don't handle data issues, i.e., PDu not at the last assessment. Using two events, PDu assessments which are not the last assessment are ignored. Using one event, PDu assessments which are not the last assessment are considered as SD. So both approaches produce incorrect results.
However, I don't think that we need to handle data issues at this point. If there is the need to identify and correct such issues, it should be done before the data is used, i.e., when the SDTMs are created or when they are read in.

For both BOR and CBOR,If **PDu remains the last recorded assessment** without follow-up confirmation, it is classified as **SD** in this vignette. However, as a more conservative approach, you may choose to classify it as **PD**. Please refer to your study documentation and protocol requirements to confirm the preferred approach to handling unresolved PDu.

```{r}
pdu_to_sd_event <- event(
  description = "Transition PDu to SD if it is the last unresolved assessment",
  dataset_name = "adrs",
  condition = PARAMCD == "OVRLRESC" & AVALC == "PDu" & LAST_REC_FL == "Y", # PDu → SD
  set_values_to = exprs(AVALC = "SD")
)

```

### Define events for Best Overall Response (BOR)

```{r, eval=TRUE, include=TRUE, message=FALSE}

bor_cr <- event(
  description = "Complete Response (CR) ",
  dataset_name = "adrs",
  condition = AVALC == "CR",
  set_values_to = exprs(AVALC = "CR")
)

bor_pr <- event(
  description = "Partial Response (PR) ",
  dataset_name = "adrs",
  condition = AVALC == "PR",
  set_values_to = exprs(AVALC = "PR")
)

> bor_non_cr_non_pd <- event(

Author: updated

)

# Confirmed PR Event with 28-day persistence
cbor_pr <- event_joined(
  description = "Confirmed Partial Response (PR)",
  dataset_name = "adrs",
  join_vars = exprs(AVALC, ADT),
  join_type = "after",
  first_cond_upper = AVALC.join %in% c("CR", "PR") & ADT.join >= ADT + 28, # Include CR as confirmation
  condition = AVALC == "PR" & all(AVALC.join %in% c("CR", "PR")), # Ensure all events align
  set_values_to = exprs(AVALC = "PR")
)
```

```{r, eval=TRUE, include=TRUE, message=FALSE}
adrs <- adrs %>%
  derive_extreme_event(
    by_vars = exprs(USUBJID),                 
    events = list(cbor_cr, cbor_pr,bor_sd,pdu_to_sd_event,bor_pd,bor_non_cr_non_pd,bor_ne,bor_ned,no_assessment_event),         
    source_datasets = list(adrs = adrs %>% filter(PARAMCD == "OVRLRESC")),
    tmp_event_nr_var = event_nr,
    order = exprs(event_nr, ADT),
    mode = "first",
    set_values_to = exprs(
      PARAMCD = "CBOR",
      PARAM = "Confirmed Best Overall Response",
      PARAMN = 6
    )
  )
```

```{r, echo=FALSE}
dataset_vignette(
  adrs %>%
    filter(PARAMCD == "CBOR"),
  display_vars = exprs(USUBJID, PARAM, PARAMCD, AVISIT, AVISITN, ADT, AVALC, AVAL)
)
```

> # Other Endpoints {#other}

Author: updated

Metastatic prostate cancer response cannot be fully captured by RECIST 1.1 criteria alone. Therefore, the PCWG3 guidelines extend these criteria to include composite response combining RECIST 1.1 criteria on soft tissue lesions and PCWG3 rules on bone bone lesions.

Note that only the PCWG3-specific steps are covered in this vignette. For extended guidance on all steps in ADRS creation, refer to the examples in [Creating ADRS (Including Non-standard Endpoints)](adrs.html).

## PCWG3 Guidelines for Prostate Cancer Progression

In metastatic prostate cancer clinical trials, efficacy is assessed using PCWG3-modified RECIST 1.1 response, PSA tumor marker results, and bone lesion assessments. Bone lesions are particularly critical for evaluating progressive disease (PD) using 2+2 rule.

### PSA Exclusion
Prostate-specific antigen (PSA) tumor marker assessments are a key component of the PCWG3 criteria for evaluating prostate cancer progression. However, they are excluded in this vignette to focus solely on deriving PCWG3-modified RECIST 1.1 endpoints related to soft tissue and bone lesions.

## Response Categories Based on PCWG3 Guidelines:

> In this vignette, we have not included RECIST 1.1 responses (soft tissue responses), as they are well-documented and widely known, and have focused solely on PCWG3 guidelines for evaluating bone scan responses in prostate cancer. PCWG3 provides specific criteria for assessing bone lesions, including categories like PDu, NED, and PD, along with the 2+2 rule for confirming progression.

Reviewer 3: @kiran-peddamudium , I can't see the update. Did you forget to push?

### 1.Progressive Disease Unconfirmed (PDu)  
-2 new bone lesions have appeared within the flare window compared to baseline.
                                       `OR`
-2 new bone lesions have appeared outside the flare window compared to the first post-treatment scan.
Note: if there is no following visit (final visit), the time point remains at PDu.

### 2.Progressive Disease (PD)  
-If 2 new bone lesions had appeared within the flare window compared to baseline (PDu), and 2 additional new
bone lesions have been found on the next scan confirming progression (PD).
-The date of progression is the date of the scan showing the first 2 lesions.
                                        `OR`
-If 2 new bone lesions had appeared outside the flare window compared to the first post-treatment scan (PDu),
and the 2 lesions are persistent on the next scan confirming progression (PD).
-The date of progression is the data of the scan that first documents the second lesion.See [**Bone Lesion Confirmation Guidance (The 2+2 Rule)**](#2+2rule) for further details.

### 3.Not Evaluable (NE)  

> -When imaging is entirely missing or was not done, but bone lesions were present at baseline.

Reviewer 2: @kiran-peddamudium remove the last sentence "but bone lesions were present at baseline"

```{r}

pdu_to_sd_event <- event(
  description = "Transition PDu to SD if it is the last unresolved assessment",
  dataset_name = "adrs",
  condition = PARAMCD == "OVRLRESC" & AVALC == "PDu" & LAST_REC_FL =="Y", # PDu → SD
  set_values_to = exprs(AVALC = "SD")
)

pdu_to_pd_event <- event_joined(
  description = "Transition PDu to PD if progression is confirmed by subsequent PD",
  dataset_name = "adrs",
  join_vars = exprs(AVALC, ADT),            
  join_type = "after",                      # Look at subsequent assessments after current record
  first_cond_upper = AVALC.join == "PD" & ADT.join > ADT, # Next record confirms PD with later date
  condition = PARAMCD == "OVRLRESC" & AVALC == "PDu",     # Current record is PDu
  set_values_to = exprs(AVALC = "PD")                     # Transition PDu → PD

> )

Reviewer 2: Not clear to me if this PDu on last assessment is converted to  -> SD is  converted in actual derived OVRLRESPC record avalc? can this be handled directly on the BOR?

bor_pr <- event(
  description = "Partial Response (PR) ",
  dataset_name = "adrs", 
  condition = AVALC == "PR",
  set_values_to = exprs(AVALC = "PR")
)

bor_non_cr_non_pd <- event(
  description = "Non-CR/Non-PD ",
  dataset_name = "adrs", 
  condition = AVALC == "Non-CR/Non-PD",
  set_values_to = exprs(AVALC = "Non-CR/Non-PD")
)

bor_sd <- event(
  description = "Stable Disease (SD) ",
  dataset_name = "adrs",
  condition = AVALC == "SD",
  set_values_to = exprs(AVALC = "SD")
)

bor_pd <- event(
  description = "Progressive Disease (PD) ",
  dataset_name = "adrs",
  condition = AVALC == "PD",
  set_values_to = exprs(AVALC = "PD")
)

bor_ne <- event(
  description = "Not Evaluable (NE) ",
  dataset_name = "adrs",
  condition = AVALC == "NE",
  set_values_to = exprs(AVALC = "NE")
)

no_assessment_event <- event(
  description = "No assessment data available for subject",
  dataset_name = "adrs",
  condition = is.na(AVALC),  # Condition where AVALC is missing
  set_values_to = exprs(AVALC = "NE")  # Assign default value as NE
)
```

## Derive Best Overall Response (BOR){#cbor}

Use the defined events to derive BOR based on the first occurrence in the adrs dataset, prioritizing 

> responses hierarchically (CR > PR > SD > PD > NE).

Reviewer 2: Above you added Non-CR/Non-PD, but here on 475 visit you haven't

Progression is confirmed based on persistence of 2 new lesions, and the progression date is recorded as the date of the first scan noting the 2 new lesions (18 weeks post-baseline).

Please check [Responses & Bone Lesion Confirmation Guidance (The 2+2 Flare Rule)](https://4870646.fs1.hubspotusercontent-na1.net/hubfs/4870646/KEO149_Info_PCWG3_03.2.pdf) for more details.

# Required Packages

The examples of this vignette require the following packages.

```{r, warning=FALSE, message=FALSE}
library(admiral)
library(admiralonco)
library(pharmaversesdtm)
library(pharmaverseadam)
library(dplyr)
library(tibble)
```

# Programming Workflow

-   [Read in Data](#readdata)
-   [Pre-processing of Input Records](#input)
-   [Derive Best Overall Response (BOR)](#bor)
-   [Derive Confirmed Best Overall Response (CBOR)](#cbor)
-   [Other Endpoints](#other)

## Read in Data {#readdata}

To begin, all data frames needed for the creation of `ADRS` should be read into the environment. This will be a company specific process. Some of the data frames needed are `ADSL` and `RS`.

> For demonstration purpose, the SDTM and ADaM datasets (based on CDISC Pilot test data) from `{pharmaversesdtm}` and `{pharmaverseadam}` are used.

Author: Updated

> For demonstration purpose, the SDTM and ADaM datasets (based on CDISC Pilot test data) from `{pharmaversesdtm}` and `{pharmaverseadam}` are used.

Author: Can I keep the code for data issue fix here , because if someone want to use rs dataset, when they see this they can see the code

> For demonstration purpose, the SDTM and ADaM datasets (based on CDISC Pilot test data) from `{pharmaversesdtm}` and `{pharmaverseadam}` are used.

Reviewer 3: Yes, good idea!

In this vignette, the `RS` SDTM dataset is expected to contain:

- Soft tissue lesion responses evaluated using RECIST 1.1 criteria at each timepoint.
- Bone lesion responses assessed across visits using PCWG3 rules.
- Combined responses derived from both RECIST 1.1 (soft tissue) and PCWG3 (bone lesion) guidelines as collected at source.

```{r message=FALSE}
# PCWG3 SDTM data
rs <- pharmaversesdtm::rs_onco_pcwg3
rs <- convert_blanks_to_na(rs)

# Exclude PSA records
rs <- rs %>% filter(RSTEST != "Tumor Marker Response")

# ADaM data
adsl <- pharmaverseadam::adsl
```

```{r, eval=TRUE, echo=FALSE}

> dataset_vignette(

Author: I have updated this for data issue, but do we have to keep this code chunk hidden? because it is inside code block echo=FALSE, I think it should be visible because if some one tries to work on this dataset, Might not know what was the issue

  "CR", "Non-PD, PDu, or NE", "PR (1)",
  "CR", "Non-PD, PDu, or NE", "Non-CR/Non-PD (2)",
  "CR", "NED", "CR"
)

overall_tpr_table %>%
  gt() %>%
  tab_header(
    title = "Table 1: Overall Time Point Response",
    subtitle = "Soft Tissue (RECIST 1.1) TPR, Bone Lesion (PCWG3) TPR, and PCWG Combined TPR"
  ) %>%
  cols_label(
    `Soft Tissue (RECIST 1.1) TPR` = "Soft Tissue (RECIST 1.1)",
    `Bone Lesion (PCWG3) TPR` = "Bone Lesion (PCWG3)",
    `Overall PCWG TPR` = "Overall PCWG"
  ) %>%
  tab_footnote(
    footnote = "* When no target and non-target lesions are identified at baseline, and no new lesions are identified on-study, the response will be No Evidence of Disease (NED)."
  ) %>%
  tab_footnote(
    footnote = "** Progressive Disease Unconfirmed (PDu): Temporary marker of possible PD where at least 2 new bone lesions are present, but an additional scan is required for confirmation."
  ) %>%
  tab_footnote(
    footnote = "(1) The overall TPR will be PR if target lesions were present at screening."
  ) %>%
  tab_footnote(
    footnote = "(2) The overall TPR will be Non-CR/Non-PD if no target lesions were present at screening."
  )
```

> **Reassigning BONERESP to PD/Non-PD**

Reviewer 2: @kiran-peddamudium @bundfussr I think this section is not required as we will update RS above as Stefan suggested

Author: Removed

output: 
  rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Creating ADRS with Prostate Cancer Working Group 3 (PCWG3) Criteria}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
library(admiraldev)
library(gt)
```

# Introduction

This article describes creating an `ADRS` ADaM dataset for prostate cancer studies based on [**Prostate Cancer Working Group 3 (PCWG3)** criteria](https://doi.org/10.1200/JCO.2015.64.2702). Most of the endpoints are derived by calling `admiral::derive_extreme_event()`.

Metastatic prostate cancer response cannot be fully captured by RECIST 1.1 criteria alone. Therefore, the PCWG3 guidelines extend these criteria to include composite response combining RECIST 1.1 criteria on soft tissue lesions and PCWG3 rules on bone bone lesions.

Note that only the PCWG3-specific steps are covered in this vignette. For extended guidance on all steps in ADRS creation, refer to the examples in [Creating ADRS (Including Non-standard Endpoints)](adrs.html).

# PCWG3 Guidelines for Prostate Cancer Response

> In metastatic prostate cancer clinical trials, efficacy is assessed using PCWG3-modified RECIST 1.1 response, PSA tumor marker results.In this vignette, we have not included RECIST 1.1 responses (soft tissue responses), as they are well-documented and widely known, and have focused solely on PCWG3 guidelines for evaluating bone scan responses in prostate cancer.

Reviewer: ```suggestion
In metastatic prostate cancer clinical trials, efficacy is assessed using PCWG3-modified RECIST 1.1 response and PSA tumor marker results. In this vignette, we have not discussed RECIST 1.1 responses (soft tissue responses), as they are well-documented and widely known, and have focused solely on PCWG3 guidelines for evaluating bone scan responses in prostate cancer.
```

Author: updated

overall_tpr_table <- tribble(
  ~`Soft Tissue (RECIST 1.1) TPR`, ~`Bone Lesion (PCWG3) TPR`, ~`Overall PCWG TPR`,
  "PD", "Any", "PD",
  "Any", "PD", "PD",
  "NE", "Non-PD, PDu, NED or NE", "NE",
  "NED", "Non-PD", "Non-CR/Non-PD",
  "NED", "PDu", "PDu",
  "NED", "NED", "NE",
  "NED", "NE", "NE",
  "SD", "Non-PD, PDu, NED or NE", "SD",
  "Non-CR/Non-PD", "Non-PD, PDu, NED or NE", "Non-CR/Non-PD",
  "PR", "Non-PD, PDu, NED or NE", "PR",
  "CR", "Non-PD, PDu, or NE", "PR (1)",
  "CR", "Non-PD, PDu, or NE", "Non-CR/Non-PD (2)",
  "CR", "NED", "CR"
)

overall_tpr_table %>%
  gt() %>%
  tab_header(
    title = "Table 1: Overall Time Point Response",
    subtitle = "Soft Tissue (RECIST 1.1) TPR, Bone Lesion (PCWG3) TPR, and PCWG Combined TPR"
  ) %>%
  cols_label(
    `Soft Tissue (RECIST 1.1) TPR` = "Soft Tissue (RECIST 1.1)",
    `Bone Lesion (PCWG3) TPR` = "Bone Lesion (PCWG3)",
    `Overall PCWG TPR` = "Overall PCWG"
  ) %>%
  tab_footnote(
    footnote = "* When no target and non-target lesions are identified at baseline, and no new lesions are identified on-study, the response will be No Evidence of Disease (NED)."
  ) %>%
  tab_footnote(

>     footnote = "** Progressive Disease Unconfirmed (PDu): Temporary marker of possible PD where at least 2 new bone lesions are present, but an additional scan is required for confirmation."

Reviewer: ```suggestion
    footnote = "** Progressive Disease Unconfirmed (PDu): Temporary marker of possible PD where at least 2 new bone lesions are present, but an additional scan is required for confirmation. To be updated to PD or non-PD once a subsequent scan is available. If this is the final visit, the response remains as PDu."
```

Author: updated

  ) %>%
  tab_footnote(
    footnote = "(1) The overall TPR will be PR if target lesions were present at screening."
  ) %>%
  tab_footnote(
    footnote = "(2) The overall TPR will be Non-CR/Non-PD if no target lesions were present at screening."
  )
```

#### Derive Combined Overall Time Point Response by Investigator (`OVRLRESC`) Records referenced from above table.

Note:`Non-PD` is changed to `NON-PD` at programming level to match with existing data.

For Scenario 11, in this vignette, it is assumed that all subjects have target lesions identified at screening. If there are subjects without target lesions identified at screening the overall response must be evaluated differently, as described in Scenario 11 and referenced in the commented code. Please review your study data to verify whether screening lesions are categorized as target or non-target, as this classification impacts the derivation of the overall response.

```{r eval=TRUE, message=FALSE, include=TRUE}
adrs <- derive_param_computed(
  dataset = adrs,

>   by_vars = exprs(!!!get_admiral_option("subject_keys"), RSEVAL, ADT, ADTF, AVISIT, AVISITN),

Reviewer 3: ```suggestion
  by_vars = exprs(!!!get_admiral_option("subject_keys"), !!!adsl_vars, RSEVAL, ADT, ADTF, AVISIT, AVISITN),
```
I'm not sure if we should add `VISIT` and `VISITNUM` as well.

      # Scenario 10: Soft Tissue = PR + Bone Lesion = NON-PD, PDu, NED, or NE -> Overall response = PR
      AVALC.SFTSRESP == "PR" & AVALC.BONERESP %in% c("NON-PD", "PDu", "NED", "NE") ~ "PR",

      # Scenario 11: Soft Tissue = CR + Bone Lesion = NON-PD, PDu, NE -> Overall response = PR
      # ((1) The overall TPR will be PR if target lesions were present at screening.)
      AVALC.SFTSRESP == "CR" & AVALC.BONERESP %in% c("NON-PD", "PDu", "NE") ~ "PR",

      # Soft Tissue = CR + Bone Lesion = NON-PD, PDu, NE -> Overall response =Non-CR/NON-PD
      # (2) The overall TPR will be Non-CR/NON-PD if no target lesions were present at screening.)
      # AVALC.SFTSRESP == "CR" & AVALC.BONERESP %in% c("NON-PD", "PDu", "NE") ~ "Non-CR/NON-PD",

      # Scenario 12: Soft Tissue = CR + Bone Lesion = NED -> Overall response = CR
      AVALC.SFTSRESP == "CR" & AVALC.BONERESP == "NED" ~ "CR",

      # Default: If conditions are not met, assign NA
      TRUE ~ NA_character_
    ),
    PARAMCD = "OVRLRESC",
    PARAM = "Overall Tumor Response by Investigator - Derived",
    PARAMN = 4,
    PARCAT1 = "PCWG3 and RECIST 1.1"
  )
)
```

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(

>   adrs,

Reviewer: ```suggestion
    adrs %>%
    arrange(!!!get_admiral_option("subject_keys"), AVISITN, PARAMN, ),
```

  dataset_name = "adrs",
  condition = AVALC == "PD",
  set_values_to = exprs(AVALC = "PD")
)

bor_ne <- event(
  description = "Not Evaluable (NE) ",
  dataset_name = "adrs",
  condition = AVALC == "NE",
  set_values_to = exprs(AVALC = "NE")
)
bor_ned <- event(
  description = "No Evidence of Disease (NED) ",
  dataset_name = "adrs",
  condition = AVALC == "NED",
  set_values_to = exprs(AVALC = "NED")
)

no_data_missing <- event(
  description = paste(
    "Define missing response (MISSING) for all patients in adsl (should be used",
    "as last event)"
  ),
  dataset_name = "adsl",
  condition = TRUE,
  set_values_to = exprs(AVALC = "MISSING"),
  keep_source_vars = adsl_vars
)

```

## Derive Best Overall Response (BOR) {#bor}

Use the defined events to derive BOR based on the first occurrence in the `adrs` dataset, prioritizing responses hierarchically (CR \> PR \> SD \> Non-CR/NON-PD \> PD \> NE \> NED \> MISSING).

In this part of the vignette, we will derive Best Overall Response based on combined response (`PARAMCD = "OVRLRESC"`) as derived above.

```{r, eval=TRUE, include=TRUE, message=FALSE}

# Derive BOR
adrs <- adrs %>%
  derive_extreme_event(

>     by_vars = exprs(USUBJID), 

Reviewer 3: ```suggestion
    by_vars = get_admiral_option("subject_keys"),
```

    events = list(bor_cr, bor_pr,bor_sd,pdu_to_sd_event,bor_pd,bor_non_crpd,bor_ne,bor_ned,no_data_missing),

>      source_datasets = list(adls = adsl, adrs = adrs %>%filter(PARAMCD == "OVRLRESC")),# Use derived responses (OVRLRESC)

Reviewer 3: Sorry, my bad. I had a typo in my suggestion.

```{r, eval=TRUE, include=TRUE, message=FALSE}
# Confirmed CR Event with 28-day persistence
cbor_cr <- event_joined(
  description = "Confirmed Complete Response (CR)",
  dataset_name = "adrs",
  join_vars = exprs(AVALC, ADT),
  join_type = "after",
  first_cond_upper = AVALC.join == "CR" & ADT.join >= ADT + 28, # Follow-up within 28-day window
  condition = AVALC == "CR" & all(AVALC.join == "CR"), # All linked records must also be CR
  set_values_to = exprs(AVALC = "CR") # Set response as Confirmed CR
)

# Confirmed PR Event with 28-day persistence
cbor_pr <- event_joined(
  description = "Confirmed Partial Response (PR)",
  dataset_name = "adrs",
  join_vars = exprs(AVALC, ADT),
  join_type = "after",
  first_cond_upper = AVALC.join %in% c("CR", "PR") & ADT.join >= ADT + 28, # Include CR as confirmation
  condition = AVALC == "PR" & all(AVALC.join %in% c("CR", "PR")), # Ensure no events other than CR or PR in between
  set_values_to = exprs(AVALC = "PR")
)
```

```{r, eval=TRUE, include=TRUE, message=FALSE}
adrs <- adrs %>%
  derive_extreme_event(

>     by_vars = exprs(USUBJID),                 

Reviewer 3: ```suggestion
    by_vars = get_admiral_option("subject_keys"),
```
